### DISCUSSION TITLE
Visualizations of high dimensional (single objective) models

### DISCUSSION THREAD

**Discussion by Friedemannn** [meta-pytorch/botorch#3298] [Q&A] [answered]
Hi all,

I'm currently working with 6D and higher search spaces in botorch.
It would be great if people could share some tips and tricks they learned of how to visualize resulting high-D models. Or maybe share links to guides they found helpful.

### My current solution:
I create 2D contour plots of each possible 6D combination (so 15). 
Each dimension axis is linearly spaced into 101 points. This means to create one plot 101^2 6D points are evaluated by the model, where only the two plotted dimensions vary, the other four are set to their optimal value found by the bo loop.

This works well to visualize (non)correlations between two parameters if the rest of the "system" is at its optimum and of course show which parameter values are feasible and which aren't.

### Projections:
I'd also like to be able to show 2D projections of the 6D space. So I'd need to average over the objective values of the other four dimensions instead of only evaluating the other 4 dimensions at their optimal value.
This would result in 15 loops with 101^2 subloops each evaluating 101^4 6D points to create the averages needed for the contour plot (If I want to keep the same resolution).

I can of course do this in parallel but this still seems like a very inefficient way of going about this with a lot of recomputation. But handling 10^6 6D points and their objective values would be impossible. 

After some LLM aided thinking I also came up with these alternatives:

- Instead of sampling over a grid with a set resolution of 101 per axis I could use a botorch provided sampler. To maybe reduce the needed samples and still give a good idea of model (average) behavior.
- I could create 15 2D wrapper models for the 6D model by running another bo loop where the objective value then is the average over the four projected dimensions.

Does anyone know a better way?

### Other ideas:
I'm also open to try non projection visualizations that you find helpful.

For example I could ditch the projections and extend my original visualization approach by also doing plots where the 4 non varying dimensions are maybe set to their median values instead of their optimal values or something similar...

**Comment by ethantenison** [comment]
@Friedemannn One thing that helped me in high-D problems is switching from conditional slices to marginal views using a partial dependence plot (PDP) matrix. I would avoid projections because they make interpretation very confusing and hard to explain to non-technical stakeholders. 

**What you’re currently doing:**
- For each 2D pair (x_i, x_j), you fix the other 4 dims at a single point (the BO optimum)
- Then evaluate the model on a grid

That answers:
> “What does the model do in this specific configuration of the system?”

This is a conditional slice. 

**What a PDP matrix does differently**
A PDP answers a different question:
> “On average, how do x_i, x_j affect the model across the distribution of the other variables?”

**Key implementation detail (this is the important part):**

At each (x_i, x_j):
- Take your dataset X (size N \times 6)
- Replace columns i, j with the grid values
- Keep the other 4 dimensions exactly as they appear in the data
- Run the model on all N modified rows
- Average the predictions

So instead of:
- 1 evaluation per grid point (your method)

you do:
- N evaluations per grid point, but only over observed configurations

**PDP:**
- Integrates over the empirical distribution
- Gives a global view of feature effects
- Is much more stable and representative

**Cost becomes:**
- 15 plots × 101^2 grid × N samples

<img width="1400" height="1400" alt="sphx_glr_partial-dependence-plot-with-categorical_002" src="https://github.com/user-attachments/assets/a245c39f-87b2-442c-8f47-7327452c611c" />

One extension is to overlay a scatter plot of the training data and the botorch suggested points.

**Reply by Friedemannn** [reply]
thanks for the suggestion!
Ok so instead of doing 101^4 samples at each 2d point we just do a greatly reduced amount of samples.
I'll try that first.

**Comment by Icheb3891** [comment]
Why not just use the derivative equations to inherite curvature of a space as you layer dimensions? This allows you to climb into higher dimensional geometry easily. Anything "linear" is confusing the computer especially as you pass the 5th dimensional grade. However a 6 dimensional model is unstable. This is literally the precise significance of prime numbers, they are the stable indices for dimensional layering.

**Comment by Icheb3891** [comment]
Also a coordinate in space doesn't exist. A singular point is exactly that, a singularity. For a point to actually exist, all inherited curvature must be undone and when that happens, you've created a black hole i.e. an asymptote within your model and unless your model can account for curvature inversion then you're on a hard path to tread.

**Reply by hvarfner** [reply, marked as answer]
@Friedemannn @ethantenison We use a couple of techniques for it, one being the 2D contour while fixing the rest to some status quo. The other one would be simple cross-validation, where you plot predicted value on the y-axis (including uncertainty) and actual value on the X-axis. This obviously doesn't give you an idea about the surface, but it can tell you a lot about the model quality.

Personally, I would recommend doing this on sampled test data too, if you are optimizing a synthetic function. What do you want to accomplish with your analysis?

**Reply by Friedemannn** [reply]
@hvarfner thanks for the suggestion!
I'll try the cross-validation plot you mentioned that sounds like a good "meta" plot to compare different runs.

We are optimizing a laser plasma accelerator using particle in cell simulation. The parameters we optimize are things like gas density, laser strength, laser focus position ...
Visualizations should show how parameters correlate and give an intuitive overview of the parameter space and where valid physical operation conditions are (not).
